glommio/src/read_write_lock.rs

```diff
@@ -0,0 +1,1214 @@
+// Unless explicitly stated otherwise all files in this repository are licensed under the
+// MIT/Apache-2.0 License, at your convenience
+
+//! Read-write locks.
+//!
+//! Provides functionality similar to the ['std::sync::RwLock'] except that lock can not be poisoned.
+//!
+//! # Examples
+//!
+//! ```
+//! use glommio::read_write_lock::ReadWriteLock;
+//! use glommio::LocalExecutor;
+//! let lock = ReadWriteLock::new(5);
+//! let ex = LocalExecutor::make_default();
+//!
+//! ex.run( async move {
+//!    // many reader locks can be held at once
+//!    {
+//!        let r1 = lock.read().await.unwrap();
+//!        let r2 = lock.read().await.unwrap();
+//!        assert_eq!(*r1, 5);
+//!        assert_eq!(*r2, 5);
+//!    } // read locks are dropped at this point
+//!
+//!    // only one write lock may be held, however
+//!    {
+//!        let mut w = lock.write().await.unwrap();
+//!        *w += 1;
+//!        assert_eq!(*w, 6);
+//!    } // write lock is dropped here
+//! });
+//! ```
+//!
+use alloc::rc::Rc;
+use core::fmt::Debug;
+use core::marker::PhantomData;
+use std::cell::{RefCell, UnsafeCell};
+use std::collections::hash_map::Entry::{Occupied, Vacant};
+use std::collections::{HashMap, VecDeque};
+use std::error::Error;
+use std::fmt::{Display, Formatter};
+use std::future::Future;
+use std::ops::{Deref, DerefMut};
+use std::pin::Pin;
+use std::task::{Context, Poll, Waker};
+
+#[derive(Clone, Copy, Eq, PartialEq, Hash, Debug)]
+struct WaiterId(u64);
+
+/// A type alias for the result of a lock method which can be blocked.
+pub type ReadWriteLockResult<T> = Result<T, LockClosedError>;
+
+/// A type alias for the result of a nonblocking locking method.
+pub type TryReadWriteLockResult<T> = Result<T, TryReadWriteLockError>;
+
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+enum WaiterKind {
+    READER,
+    WRITER,
+}
+
+///Error which indicates that ReadWriteLock is closed and can not be use
+///to request lock access
+pub struct LockClosedError;
+
+impl Error for LockClosedError {}
+
+impl Display for LockClosedError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
+        f.write_str("Lock already closed")
+    }
+}
+
+impl Debug for LockClosedError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
+        f.write_str("Closed")
+    }
+}
+
+///Error which indicates reason of the error returned by [`try_read`] and ['try_write'] methods
+///on ['ReadWriteLock']
+pub enum TryReadWriteLockError {
```

Comment:
what's the advantage of using our own error class?
For the channels I have used standard `io::Error`. A closed lock can return `BrokenPipe` and also `WouldBlock`.
The advantage of using `io::Error` is that we can just convert easily between errors coming from different places of the system.

Reply:
My gut feeling is that we should use custom error for channels as well, but I can't really articulate this without reflecting on this for like half an hour :)

Reply:
I noticed that you use io::Error, but errors in "sync" API, let's start to call it like this, have nothing to do with io. It is not so hard to implement From trait for them. But when I look at "Broken pipe" message in io::Error I really do not understand how that is related to the problem and it makes me fell that developers decided to do not put too much of effort to create "clear" API. That is why I used a different error class. It is more about the perception of API than functionality. Do you have a strong feeling that io::Error should be used instead ?

Reply:
>It is more about the perception of API than functionality.

It's a bit less nebulous than that. `io::Error` has support for dynamically dispatched errors, it is a bit like `RuntimeException` from other languages, and is the perfect choice when:

* you want to expose underlying OS-errors as is
* you want some generic error with type erasure 

(I've wrote a bit about its design here: https://matklad.github.io/2020/10/15/study-of-std-io-error.html)

For sync primitives, the set of error conditions is fixed (closed), so an `enum` works better. And, as an argument from authority, crossbeam-channel does that.

Reply:
I have been welcoming @matklad 's perspective on this as a more experienced rustacean and he seems to slightly favor what you did.

For me personally I always found "io" to be a bit of a misnomer: they are errors returned in system calls, and not all system calls do I/O anyway...

The appeal of using `io::Result` for me is that you can call that together in a function that does other things as well and because they all return the same error you can `?` without conversions.

You can always implement conversions, that's true, but as far as I know you cannot implement out-of-crate conversions. So the poor user of this would not be able to easily convert to `io::Error`: we'd have to provide the conversion class.

And if we are the ones providing the conversion class, why not go all the way ?

Unless I am missing something with the out-of-crate conversion, then if you keep this as a standard error I think we should at least provide the conversion to `io::Error`. `io::Error`s are so ubiquitous in an I/O-intensive crate, which is what most functions end up returning anyway. (Think of using a `rwlock` to intermediate access to a file... you will want your function that does that to return `io::Error`!)

Reply:
I am fine with custom errors but in that case I think we should implement our own conversions to `io::Error`. That is, unless I am mistaken about out-of-crate implementation of traits.

Think of the example I gave above, where you may have a `read_file` function that protects the file with a `rwlock`.
It's very clear to me that the top level function should return `io::Result` so we should make as easy as possible to convert back to that.

Reply:
Hi, @glommer @matklad let me summarize. So we keep custom errors but will provide conversion to the io::Error is that correct? @matklad thank you for the link I will read it today. BTW if we use Lock errors for the RwLock it would be also useful to use the same error for the Semaphore for the consistency fo the view. Do not you think so?

Reply:
I am ok with using custom errors with a provided conversion.

Indeed if you are willing to go the extra mile, converting the existing errors for Semaphore and even Channels for consistency would be super.

Reply:
Sure will do. Will provide a separate PRs.